pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiGroups"), rule.APIGroups, "if '*' is present, must not specify other api groups"))
+	}
+	if hasWildcard(rule.Resources) && len(rule.Resources) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resources"), rule.Resources, "if '*' is present, must not specify other resources"))
+	}
+
+	// only supported verbs are allowed
+	if !supportedVerbs.IsSuperset(sets.NewString(rule.Verbs...)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("verbs"), rule.Verbs, supportedVerbs.List()))
+	}
+
+	return allErrs
+}
+
+// ValidateFlowSchemaStatus validates status for the flow-schema.
+func ValidateFlowSchemaStatus(status *flowcontrol.FlowSchemaStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// conditions will not be validated
+	return allErrs
+}
+
+// ValidatePriorityLevelConfiguration validates priority-level-configuration.
+func ValidatePriorityLevelConfiguration(pl *flowcontrol.PriorityLevelConfiguration) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&pl.ObjectMeta, false, ValidatePriorityLevelConfigurationName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationSpec(&pl.Spec, pl.Name, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationStatus(&pl.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidatePriorityLevelConfigurationSpec validates priority-level-configuration's spec.
+func ValidatePriorityLevelConfigurationSpec(spec *flowcontrol.PriorityLevelConfigurationSpec, name string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch spec.Type {
+	case flowcontrol.PriorityLevelQueuingTypeExempt:
+		if name != flowcontrol.PriorityLevelConfigurationNameExempt {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("type"), fmt.Sprintf(`must be named "exempt" to be exempt queuing type`)))
+		}
+	case flowcontrol.PriorityLevelQueuingTypeQueueing:
+		allErrs = append(allErrs, ValidatePriorityLevelQueuingConfiguration(spec.Queuing, fldPath.Child("queuing"))...)
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("type"), spec.Type, supportedQueuingType.List()))
+	}
+	return allErrs
+}
+
+// ValidatePriorityLevelQueuingConfiguration validates queuing-configuration for a priority-level
+func ValidatePriorityLevelQueuingConfiguration(queuing *flowcontrol.QueuingConfiguration, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if queuing.AssuredConcurrencyShares < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("assuredConcurrencyShares"), queuing.AssuredConcurrencyShares, "must be positive"))
+	}
+	if queuing.QueueLengthLimit < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("queueLengthLimit"), queuing.QueueLengthLimit, "must be positive"))
+	}
+
+	// validate input arguments for shuffle-sharding
+	if queuing.Queues < 0 {
```

Comment:
<=

Reply:
(or fix the error message if 0 is a valid number of queues?)

Reply:
> yes, the defaulting will make sure 0 wont happen, are we validating that anyway? @lavalamp